File: least-common-multiple.py
"""
LCM Calculator – Least Common Multiple of Two or More Integers

This script calculates the Least Common Multiple (LCM) of two or more integers
using the classic Euclidean algorithm. Accepts space- or comma-separated input.

Author: Scott Douglass
Date: 2025-04-20
"""

from functools import reduce

def gcd(a, b):
    """Compute the GCD of two integers using the Euclidean algorithm."""
    while b != 0:
        a, b = b, a % b
    return abs(a)

def lcm(a, b):
    """Compute the LCM of two integers."""
    return abs(a * b) // gcd(a, b) if a and b else 0

def lcm_list(numbers):
    """Compute the LCM of a list of integers."""
    return reduce(lcm, numbers)

def parse_input(raw):
    """
    Parses user input allowing both space- and comma-separated numbers.

    Args:
        raw (str): Raw input string

    Returns:
        list[int]: List of integers
    """
    cleaned = raw.replace(',', ' ')
    try:
        return list(map(int, cleaned.split()))
    except ValueError:
        print("❌ Invalid input. Please enter integers separated by space or commas.")
        exit(1)

if __name__ == "__main__":
    input_str = input("Enter integers (space or comma separated, default: 12 18 30): ").strip()
    numbers = parse_input(input_str) if input_str else [12, 18, 30]
    result = lcm_list(numbers)
    print(f"LCM{tuple(numbers)} = {result}")

File: least-common-multiple.cpp
/*
 * Least Common Multiple Calculator (LCM)
 *
 * This program computes the Least Common Multiple (LCM) of two or more integers.
 * It uses the Euclidean algorithm to find the Greatest Common Divisor (GCD),
 * and then applies the formula: LCM(a, b) = |a * b| / GCD(a, b)
 *
 * Features:
 * - Accepts command-line input of space- or comma-separated integers.
 * - Uses standard input if no command-line arguments are provided.
 * - Validates user input and reports errors.
 *
 * Author: Scott
 * Date: 2025-07-20
 */

#include <iostream>
#include <sstream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstdlib>
#include <cctype>

int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return std::abs(a);
}

int lcm(int a, int b) {
    if (a == 0 || b == 0) return 0;
    return std::abs(a * b) / gcd(a, b);
}

int lcm_list(const std::vector<int>& nums) {
    if (nums.empty()) return 0;
    int result = nums[0];
    for (size_t i = 1; i < nums.size(); ++i) {
        result = lcm(result, nums[i]);
    }
    return result;
}

std::vector<int> parse_input(const std::string& input) {
    std::vector<int> numbers;
    std::string cleaned;
    for (char ch : input) {
        if (ch == ',') cleaned += ' ';
        else cleaned += ch;
    }

    std::istringstream iss(cleaned);
    std::string token;
    while (iss >> token) {
        try {
            numbers.push_back(std::stoi(token));
        } catch (...) {
            std::cerr << "❌ Invalid input: '" << token << "' is not an integer.\n";
            exit(1);
        }
    }
    return numbers;
}

int main() {
    std::string input;
    std::cout << "Enter integers (space or comma separated, default: 12 18 30): ";
    std::getline(std::cin, input);

    std::vector<int> numbers = input.empty() ? std::vector<int>{12, 18, 30} : parse_input(input);
    int result = lcm_list(numbers);

    std::cout << "LCM(";
    for (size_t i = 0; i < numbers.size(); ++i) {
        std::cout << numbers[i];
        if (i != numbers.size() - 1) std::cout << ", ";
    }
    std::cout << ") = " << result << std::endl;

    return 0;
}
